pkg/worker/redis.go

```diff
@@ -0,0 +1,203 @@
+	gob.Register(args)
+}
+
+func loggerWithJob(ctx context.Context, job *Job) *zerolog.Logger {
+	logger := ctxval.Logger(ctx).With().
+		Str("job_id", job.ID.String()).
+		Str("job_type", string(job.Type)).
+		Interface("job_args", job.Args).Logger()
+	return &logger
+}
+
+func (w *RedisWorker) Enqueue(ctx context.Context, job *Job) error {
+	logger := loggerWithJob(ctx, job)
+	var err error
+
+	if job.ID == uuid.Nil {
+		job.ID, err = uuid.NewRandom()
+		if err != nil {
+			return fmt.Errorf("unable to generate UUID: %w", err)
+		}
+	}
+
+	var buffer bytes.Buffer
+	enc := gob.NewEncoder(&buffer)
+	err = enc.Encode(&job)
+	if err != nil {
+		return fmt.Errorf("unable to encode args: %w", err)
+	}
+
+	cmd := w.client.LPush(ctx, w.queueName, buffer.Bytes())
+	if cmd.Err() != nil {
+		logger.Error().Err(err).Msg("Unable to push job into Redis")
+		return fmt.Errorf("unable to push job into Redis: %w", cmd.Err())
+	}
+
+	result, err := cmd.Result()
+	if err != nil {
+		return fmt.Errorf("unable to process result: %w", err)
+	}
+	logger.Info().Int64("job_result", result).Msg("Pushed job successfully")
+	return nil
+}
+
+func (w *RedisWorker) Stop(ctx context.Context) {
+	logger := ctxval.Logger(ctx)
+	close(w.closeCh)
+	logger.Info().Msg("Waiting for all workers to finish")
+	w.loopWG.Wait()
+	w.handleWG.Wait()
+	logger.Info().Msg("Done waiting for all workers to finish")
+}
+
+func (w *RedisWorker) DequeueLoop(ctx context.Context) {
+	logger := ctxval.Logger(ctx)
+	pollers := math.Min(w.pollMax, runtime.NumCPU()+1)
+	logger.Info().Msgf("Starting %d polling goroutines", pollers)
+	for i := 1; i <= pollers; i++ {
+		w.loopWG.Add(1)
+		go w.dequeueLoop(ctx)
+	}
+}
+
+//nolint:gosec
+func (w *RedisWorker) dequeueLoop(ctx context.Context) {
+	defer w.loopWG.Done()
+	logger := ctxval.Logger(ctx)
+
+	// randomize polling intervals for a better spread (random delay up to one read timeout)
+	randMs := rand.Int63n(w.pollInterval.Milliseconds())
+	logger.Info().Msgf("Worker start sleep (ms): %d", randMs)
+	time.Sleep(time.Duration(randMs) * time.Millisecond)
+
+	for {
+		select {
+		case <-w.closeCh:
+			logger.Info().Msg("Shutting down a Redis poller (stop)")
+			return
+		case <-ctx.Done():
+			logger.Info().Msg("Shutting down a Redis poller (cancel)")
+			return
+		default:
+			w.fetchJob(ctx)
+		}
+	}
+}
+
+func (w *RedisWorker) fetchJob(ctx context.Context) {
+	res, err := w.client.BLPop(ctx, w.pollInterval, w.queueName).Result()
+
+	if errors.Is(err, redis.Nil) {
+		// timeout occurred
+	} else if err != nil {
+		logger := ctxval.Logger(ctx)
+		logger.Error().Err(err).Msg("Error consuming from Redis queue")
+	} else {
+		var job Job
+		dec := gob.NewDecoder(strings.NewReader(res[1]))
+		err = dec.Decode(&job)
+		logger := loggerWithJob(ctx, &job)
+		if err != nil {
+			logger.Error().Err(err).Msg("Unable to unmarshal job payload, skipping")
+		}
+
+		w.handleWG.Add(1)
+		atomic.AddInt64(&w.inFlight, 1)
+		go w.processJob(ctx, &job)
+	}
```

Comment:
Just a nitpick, but this feels a bit more natural for me.
```suggestion
	if errors.Is(err, redis.Nil) {
		// timeout occurred
		return
	} else if err != nil {
		logger := ctxval.Logger(ctx)
		logger.Error().Err(err).Msg("Error consuming from Redis queue")
		return
	}
	var job Job
	dec := gob.NewDecoder(strings.NewReader(res[1]))
	err = dec.Decode(&job)
	logger := loggerWithJob(ctx, &job)
	if err != nil {
		logger.Error().Err(err).Msg("Unable to unmarshal job payload, skipping")
	}

	w.handleWG.Add(1)
	atomic.AddInt64(&w.inFlight, 1)
	go w.processJob(ctx, &job)
```

Reply:
Indeed, because I extracted this from the loop that was the reason it was written this ugly way.